Add defensive isMounted checks before setting state via useMutation Hook

alessbell opened the pull request:
This PR adds checks in two places inside of `useMutation` to verify that the component is mounted before setting state with `setResult`, which fixes https://github.com/apollographql/apollo-client/issues/9858. ~The code changes are trivial, and while I initially thought it might be possible to add some test coverage for this minor change, that's proven more difficult than anticipated.~ (see edit below)

## Testing

When initially calling `createTodo` in my test component's cleanup function, I was not seeing the expected React error re: setting state on an unmounted component, but I did see the error when I called `reset` after `unmount` (which was also calling `setResult` without checking the `isMounted` boolean).
 
![Screen Shot 2022-07-05 at 12 32 09 PM](https://user-images.githubusercontent.com/5139846/177403137-984b0b59-5bac-4eb0-a517-da2818bccbda.png)

~It looks like I'm at an impasse when it comes to making assertions against component state **after the component unmounts** (I want to call unmount in the middle of my test—it's also called automatically after each test via cleanup function—in order to ensure the state _hasn't_ changed), as subsequent calls to `waitForNextUpdate` time out. Would love any suggestions, I'll also ask around in the Testing Library community :-) I'd really like to have tests in place to prevent regressions, but regardless, it's a needed fix for `useMutation`.~

Edit: spying on console.error here is sufficient. I've updated the tests, which now fail if either `isMounted` check added in the PR is removed.

### Checklist:

- [x] ~If this PR is a new feature, please reference an issue where a consensus about the design was reached (not necessary for small changes)~
- [x] Make sure all of the significant new logic is covered by tests

dylanwulf commented:
> It looks like I'm at an impasse when it comes to making assertions against component state after the component unmounts (I want to call unmount in the middle of my test—it's also called automatically after each test via cleanup function—in order to ensure the state hasn't changed), as subsequent calls to waitForNextUpdate time out. Would love any suggestions

What I did in #9801 was just spy on `console.error` and assert that it never got called. Probably not the best solution, but it works!

alessbell commented:
> What I did in https://github.com/apollographql/apollo-client/pull/9801 was just spy on console.error and assert that it never got called. Probably not the best solution, but it works!

For this case, that should suffice :) Updated - thanks!

alessbell commented:
@benjamn removing either `isMounted` check this PR adds now fails the test (thanks to @[dylanwulf](https://github.com/dylanwulf)'s suggestion 👍). Updated and squashed into a single commit.

hwillson reviewed (approved):
Woohoo - this is really great @alessbell (and thanks for the spying on `console.error` suggestion @dylanwulf 🙇). A big 👍 from me!
